In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('./sonar_dataset.csv', header=None)
df.head()

,0,1,2,3,4,5,6,7,8,9,...,51,52,53,54,55,56,57,58,59,60
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032,R
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044,R
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078,R
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117,R
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094,R


In [3]:
df.shape

(208, 61)

In [4]:
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [5]:
df[60].value_counts()

60
M    111
R     97
Name: count, dtype: int64

In [6]:
x = df.drop(60, axis='columns')
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [7]:
y = pd.get_dummies(y, drop_first=True, dtype=int)
y.head()

,R
0,1
1,1
2,1
3,1
4,1


In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.25)
X_train.shape, X_test.shape

((156, 60), (52, 60))

In [9]:
import tensorflow as tf
from tensorflow import keras

In [10]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(10, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid'),
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=70, batch_size=9)

Epoch 1/70
18/18 [==============================] - 3s 10ms/step - loss: 0.6810 - accuracy: 0.5577
Epoch 2/70
18/18 [==============================] - 0s 10ms/step - loss: 0.6708 - accuracy: 0.5769
Epoch 3/70
18/18 [==============================] - 0s 10ms/step - loss: 0.6484 - accuracy: 0.5962
Epoch 4/70
18/18 [==============================] - 0s 9ms/step - loss: 0.6200 - accuracy: 0.6859
Epoch 5/70
18/18 [==============================] - 0s 10ms/step - loss: 0.6026 - accuracy: 0.7115
Epoch 6/70
18/18 [==============================] - 0s 10ms/step - loss: 0.5726 - accuracy: 0.6987
Epoch 7/70
18/18 [==============================] - 0s 11ms/step - loss: 0.5392 - accuracy: 0.7949
Epoch 8/70
18/18 [==============================] - 0s 10ms/step - loss: 0.5116 - accuracy: 0.7628
Epoch 9/70
18/18 [==============================] - 0s 10ms/step - loss: 0.4952 - accuracy: 0.7500
Epoch 10/70
18/18 [==============================] - 0s 10ms/step - loss: 0.5078 - accuracy: 0.7244
Epoch 11/7

In [11]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 7ms/step - loss: 0.3768 - accuracy: 0.8269


[0.3768300712108612, 0.8269230723381042]

In [12]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 6ms/step
[9.2098939e-01 9.9366546e-01 9.9997008e-01 2.8629017e-01 1.5680103e-04
 1.7188364e-04 6.5417774e-02 9.4749957e-01 6.7510659e-01 9.9993622e-01]
[1. 1. 1. 0. 0. 0. 0. 1. 1. 1.]


In [13]:
from sklearn.metrics import confusion_matrix, classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.80      0.83      0.82        24
           1       0.85      0.82      0.84        28

    accuracy                           0.83        52
   macro avg       0.83      0.83      0.83        52
weighted avg       0.83      0.83      0.83        52



In [16]:
modelD = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(10, activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(1, activation='sigmoid'),
])

modelD.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modelD.fit(X_train, y_train, epochs=70, batch_size=9)

Epoch 1/70
18/18 [==============================] - 1s 13ms/step - loss: 0.7006 - accuracy: 0.4615
Epoch 2/70
18/18 [==============================] - 0s 12ms/step - loss: 0.6815 - accuracy: 0.5897
Epoch 3/70
18/18 [==============================] - 0s 12ms/step - loss: 0.6824 - accuracy: 0.5577
Epoch 4/70
18/18 [==============================] - 0s 12ms/step - loss: 0.6662 - accuracy: 0.6026
Epoch 5/70
18/18 [==============================] - 0s 13ms/step - loss: 0.6931 - accuracy: 0.5192
Epoch 6/70
18/18 [==============================] - 0s 14ms/step - loss: 0.6555 - accuracy: 0.6410
Epoch 7/70
18/18 [==============================] - 0s 13ms/step - loss: 0.6798 - accuracy: 0.5897
Epoch 8/70
18/18 [==============================] - 0s 13ms/step - loss: 0.6328 - accuracy: 0.6026
Epoch 9/70
18/18 [==============================] - 0s 13ms/step - loss: 0.6521 - accuracy: 0.6154
Epoch 10/70
18/18 [==============================] - 0s 14ms/step - loss: 0.6224 - accuracy: 0.6987
Epoch 11/

In [19]:
modelD.evaluate(X_test, y_test)

2/2 [==============================] - 0s 12ms/step - loss: 0.3561 - accuracy: 0.8654


[0.35610491037368774, 0.8653846383094788]

In [20]:
y_pred = modelD.predict(X_test).reshape(-1)
print(y_pred[:10])

y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 6ms/step
[0.7506882  0.966773   0.9979165  0.13542002 0.00543614 0.00502744
 0.0211742  0.7968752  0.25758767 0.9936371 ]
[1. 1. 1. 0. 0. 0. 0. 1. 0. 1.]


In [21]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.77      1.00      0.87        24
           1       1.00      0.75      0.86        28

    accuracy                           0.87        52
   macro avg       0.89      0.88      0.86        52
weighted avg       0.90      0.87      0.86        52



## You can see that by using dropout layer test accuracy increased from 0.77 to 0.81